src/D2L.Bmx/Okta/OktaSessionStorage.cs

```diff
@@ -0,0 +1,64 @@
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using D2L.Bmx.Okta.Models;
+
+namespace D2L.Bmx.Okta;
+
+internal class OktaSessionStorage {
+
+	internal static void SaveSessions( List<OktaSessionCache> sessions ) {
+
+		string jsonString = JsonSerializer.Serialize(
+			sessions.ToArray(),
+			SourceGenerationContext.Default.OktaSessionCacheArray );
+		File.WriteAllText( SessionsFileName(), jsonString );
+	}
+
+	internal static List<OktaSessionCache> Sessions() {
+
+		var bmxDirectory = BmxDir();
+		var sessionsFileName = SessionsFileName();
+
+		if( !Directory.Exists( bmxDirectory ) ) {
+			Directory.CreateDirectory( bmxDirectory );
+		}
+
+		if( !File.Exists( sessionsFileName ) ) {
+			File.WriteAllText( sessionsFileName, "[]" );
+		}
+
+		try {
+			var sessionsJson = File.ReadAllText( SessionsFileName() );
+			var sessions = JsonSerializer.Deserialize( sessionsJson, SourceGenerationContext.Default.OktaSessionCacheArray );
+			if( sessions is not null ) {
+				return sessions.ToList();
+			}
+		} catch( JsonException ) {
+			return new();
+		}
+		return new();
+	}
+
+	internal static string SessionsFileName() {
+		return Path.Join( UserHomeDir(), ".bmx3", "sessions" );
+	}
+
+	internal static string BmxDir() {
+		return Path.Join( UserHomeDir(), ".bmx3" );
+	}
+
+	internal static string UserHomeDir() {
```

Comment:
Lets consider using SpecfialFolder's UserProfile instead https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder  

A glance through the dotnet source suggests they already put in the work to correctly get the home folder on linux, but it wouldn't doing a quick test, its not very readable code to me 😆

Reply:
Oh yea nice. I didn't know about that. Let's use that instead

Reply:
No concerns from me aside from it being a departure from current BMX which we wanted to avoid at least in this initial release.

> Oh yea nice. I didn't know about that. Let's use that instead

Best to check on a common Linux distro to make sure it finds /home/<user> properly, the code I see here is a little :suspect: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Environment.SpecialFolder.cs,7d2f1469d916fc63, not quite sure how to interpret its ability to identify home folder correctly https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs,7d2f1469d916fc63

Reply:
Yea I am not strongly for appdata. I'll check in docker where that special folder goes in linux

Reply:
I'd strongly recommend using SpecialFolder.

Not 100% on AppData, but I'd argue that if we switch, this initial release is exactly when we'd want to do it.
This will be a big release in any case and comes with a major version bump, so people will expect incompatible changes. We'd want to avoid breaking changes in the future (after the initial .NET release) without a good reason.